### ISSUE TITLE
Identity system implementation

### ISSUE THREAD

**Issue by digitalic** [space-wizards/space-station-14#105]
### Why a new system at all?

SS13's identity system sucks. There's many reasons but basically three fundamental issues.

- The face organ, when visible (and undamaged), shows the true name of the person who had that face at roundstart/when they were created. In other words, regardless of who you are, whenever you hover over someone/examine someone/see someone do something, and you can see their face, you are given the name of the person who had that face at roundstart. This makes no sense if your character doesn't actually know that person, and just leads to confusion and failed RP. One example is crewmembers recognizing nuclear operatives' names through their faces, when they have no IC reason to do so.

- The voice organ, when heard, shows the true name of the person whose voice that was at roundstart/creation. This has basically the same issues as faces. You can hear a nuke op talk over the radio, and know his name as a result.

- The face organ and voice organ are the exact same thing. This exacerbates the above problems. For example, you hear a voice of "Bob" speak from an intercom, and then later meet the face of "Bob", who doesn't say anything, yet you know that they were the same person. It's actually difficult to separate the two in SS13 - try to think of what you should see in the chatbox when someone with a mismatched face, voice, and ID says something.

You might not consider these very important issues, especially if you tend to only play on "classic" SS13 servers. These tends to be centered around the events of a station where all crew are expected to know each other, with gamemodes and events built around this fact.

But consider non-"classic" servers, such as open-world servers like fallout13, or war servers like lebensraum. These are heavily impacted by the ability to instantly recognize strangers' and enemy combatants' names through their faces and voices. Also consider custom events and gamemodes that can't make it into "classic" SS13 servers because of the defeating power of SS13's sloppy identification system.

Faces and voices need to be replaced with something better, or SS14's potential to branch into wild new situations will be forever hindered.

### Untangling the visualprints, audioprints, and audiovisualprints

Before we can work to replace faces and voices, we need to untangle SS13's system of showing them to us. Let's look at all situations where faces and voices are actually shown to us. Call a face's name a "FACEPRINT" and a voice's name a "VOICEPRINT". In SS13, these are both just the character's real name.

- When shift-clicking/examining a person, at the top of the examine message:
  'This is FACEPRINT (as OPTIONALLY_MISMATCHED_ID).'
- When hovering over a person, in the bottom left corner of the screen:
  'FACEPRINT (as OPTIONALLY_MISMATCHED_ID)'
- When seeing someone perform an action that is noted in the chat log:
  'FACEPRITN (as OPTIONALLY_MISMATCHED_ID) does X'
- You hear someone without seeing their face (like over the radio).
  'VOICEPRINT says, "X".'
- You hear someone while seeing their face (like talking in person).
  'FACEPRINT/VOICEPRINT (as OPTIONALLY_MISMATCHED_ID) says, "X".'

There's an obvious pattern here. The lines involving a way to visually identify a person have a kind of "visual fingerprint", a "visualprint": 

FACEPRINT (as OPTIONALLY_MISMATCHED_ID). 

The line involving indirect speech can be considered to have an "audio fingerprint", an "audioprint": 

VOICENAME.

The line involving face-to-face speech can be considered to have an "audiovisual fingerprint", an "audiovisualprint":

BOTH_FACE_AND_VOICE_NAME (as OPTIONALLY_MISMATCHED_ID)

This looks the same as the visualprint, because faces and voices are the same thing in SS13. We're planning on changing that, so we still need a separate "audiovisualprint" term.

Let's look at the same five examples:

- When shift-clicking/examining a person, at the top of the examine message:
  'This is VISUALPRINT.'
- When hovering over a person, in the bottom left corner of the screen:
  'VISUALPRINT'
- When seeing someone perform an action that is noted in the chat log:
  'VISUALPRINT does X'
- You hear someone without seeing their face (like over the radio).
  'AUDIOPRINT says, "X".'
- You hear someone while seeing their face (like talking in person).
  'AUDIOVISUALPRINT says, "X".'

Visualprints, audioprints, and audiovisualprints have one job: let players quickly IC identify other characters when interacting with them. The reason we labeled these new terms is because we need to fix them before working on replacing SS13's faces and voices system.

The first change we can implement is to make VISUALPRINT no longer based on a FACE_NAME and an ID_NAME at all times. Instead, it picks the single most unique element from a list of available visual identifiers. This will probably be the face and then the most unique identifying clothing. For example:

- Visualprint of someone with a shown face.
  'That's FACE_NAME.'
  'That's "Bob Jones".'
- Visualprint of someone with a hidden face and shown ID.
  'That's someone wearing ID_NAME.'
  'That's someone wearing "ID: Bob Jones".'
- Visualprint of someone with a hidden face and ID.
  'That's someone wearing NOTABLE_CLOTHING.'
  'That's someone wearing "Engineering Hardsuit".'

This can seem a bit long and verbose, but you can always just get rid of the sentence filler or make it a lighter text color, and players will catch on. E.g.

- '"ID: Bob Jones" performs the "flossing" FortNite dance. Absolutely epic! You feel like you want to dance along.'

The obvious benefit to all this is that people don't all come up as "Unknown" if hiding their face and ID, and can still be differentiated. The obvious change here though, is that IDs are no longer always implicitly included. I consider this a benefit, because a mismatched ID sticks out far too much like a sore thumb in the chatlog and examination, which hinders antags' ability to blend in, an ability that is already stretched far too thin in regular SS13. We can still bolt IDs onto the end of visualprints if we decide to, but for now we'll keep them off.

We also need to change audiovisualprints to be visualprint + audioprint.

- You hear someone while seeing their face (like talking in person).
  'VISUALPRINT, speaking as AUDIOPRINT, says, "X".'
  '"ID: Bob Jones", speaking as "Bob Jones", says, "X".'

- We can squish that if we need.
  'VISUALPRINT AUDIOPRINT says, "X".'
  '"ID: Bob Jones" "Voice: Bob Jones" says, "X".'

We don't need to do anything to the formatting of voiceprints, since they're just the voice name.

### The part where we actually fix faces and voices instead of making questionable UI changes

So faceprints and voiceprints are included in visualprints, audioprints, and audiovisualprints.

The problem is that faceprints and voiceprints are always just the real name of the person who started with that face and voice. Let's solve this issue, starting with faces.

First things first, we're stripping off the real name from faces. Now they're just "unknown".

- 'That's "unknown".'

Now the problem is we can't identify different peoples' faces. We need a way to do this.

How about a description? Have each face generate a description composed of adjectives at roundstart. You can even let players choose which adjectives describe their face out of a preset list.

'That's a "High-browed, blue-eyed, thin-lipped, narrow-nose, thick-chinned, round-eared, wide-cheeked, almond-eyelidded, long-faced, oily-skinned..." person.'

A little bit too long, I think. Let's keep the description, but cut out two random adjectives and use that as the faceprint.

'That's a "high-browed, oily-skinned" person.'

And if you want to see a face's full description, just hover over the faceprint name and it'll appear in a tooltip.

Now we have both a quick way to recognize unique faces, and a more thorough way to see all of their details. But we do have an issue if the faceprint is the same for everyone who sees it. People will just use the faceprint IC, like a name, yelling "; help high-browed rosy-cheeked is succing me in dorms".

To fix that we'll make the description-generated faceprint random for each character. Both of them will see the same "high-browed, blue-eyed, thin-lipped..." description, but their faceprints will be different. One player's "high-browed" is another player's "long-faced'. They can still call out an assailant as a high-browed rosy-cheeked douchebag, but that's very little use to others, because then they have to go through people's face descriptions trying to match that and not get a false positive. Better call out other features like race or hair, like one would IRL.

It's pointless having faceprints if they are re-generated every time you see someone, so they should stick around in your character's memory, a "face memory database". A player could access this and scroll through a sorted list of known faces and their descriptions.

If we wanted to be thorough, parts of people's face description would "decay" in memory. This represents your character forgetting face details. They could eventually even completely forget a faceprint eventually, which would make them regenerate a new faceprint when seeing that same person. I don't think going this far is necessary, though.

We can make voiceprints the same way we did for faces. Randomly/manually generate a voice description.

'You hear a "Deep, nasal, unenunciated, clipped, staccato, gruff, unappealing..." voice.'

Take three adjectives out of the bag, make a voiceprint:

'The "deep, clipped, gruff" voice says, "X"'.

Works dandy. We can have the same hover-for-full-description tooltip, and a player-accessible voice memory database with all the bells and whistles.

What does an audiovisualprint look like now?

- You hear someone while seeing their face.
  'AUDIOVISUALPRINT says, "X".'
  'The "high-browed, oval-faced" person, speaking with a "deep, clipped, gruff" voice, says "X".'
- Squished:
  '"High-browed, oval-faced" - "deep, clipped, gruff" says, "X".

Looks good. You can instantly tell different faces and voices apart. Your character's ability to remember who a face or voice belongs to matches your own: you will remember the faceprints and voiceprints of the people that your character cares about, and forget the prints of those that they don't. This solution absolutely works well for a very small server, 10 to 20 pop, where one is not expected to meet a lot of people.

### Naming, renaming, and unnaming

Still, there's more issues. For one, it's somewhat mentally taxing to translate "high-browed, rosy-cheeked" into someone's name every time. This takes too long for a round-based gamemode that can have many important things happen in short amounts of time.

Another issue is there are cases when your character is supposed to know a person, and consequently their face and voice, but you as a player don't. For example, think of your departmental coworkers on a space station. Your character should know them, but you as a player might not have seen them around before.

Both of these issues can be solved by adding a "person name" field to the face and voice organs. This field either starts off blank if the character doesn't know that face/voice, or it starts pre-filled with the name of a character if they do know that face/voice (more on pre-knowledge later). When set, the "person name" field becomes the new face/voiceprint in the UI. For example:

- You examine a nuke op, whose face you don't know at roundstart.
  'That's a "high-browed, rosy-cheeked" person.'

- You examine the HoP, whose face you do know at roundstart.
  'That's "Bob Johnson".'

You can rename faces and voices at will. One way to set up the UI to do this is to have the player click on the name, and then click on a "rename this [face/voice]" button that pops up in a panel underneath. This will turn the name into a text entry field that captures keyboard focus. If you type in a name, it will either add that as a person-name, or replace the existing one. If you leave the field blank, it will remove the person-name and you will go back to seeing the original face/voiceprint.

We might want a way to see previous renames and the original face/voiceprint. This information can be added next to the description tooltip that pops up when hovering over a name.

Players might not always use the rename feature to write names. They might use the field to add short notes next to faces and voices, such as "Bob Johnson - Traitor". This might seem a little gamey, but think of it as a character's strong mental association of a person with an important quality of theirs. This could even be done automatically at roundstart - you can start off knowing people by "Bob Johnson (Botanist)", so that you remember their role without needing to know see their ID or check the manifest. If you don't like this, you can just make a rule against it and log all renames for easy bans.

By the way, what about audiovisualprints?

- You hear the HoP talk in person.
  '"Bob Johnson", with "Bob Johnson"'s voice says, "X".

This is repetitive. If the face and voice match, we should collapse them into one.

- You hear the HoP talk in person.
  '"Bob Johnson" says, "X".

That's better. It does mess a little with renaming though. In these special instances, we will have three buttons in the name-clicked-popup-panel: "rename face", "rename voice", and "rename both face and voice". It's not very likely that someone would use the first two, but they might in esoteric situations. 

"Rename both face and voice" is actually a useful button in two other situations:

- You hear an unknown person talk to you.
  'An "high-browed, rosy-cheeked" face, with a "deep, clipped, gruff" voice says, "The name's Bob Johnson".

Click one name, click "rename both", type in Bob Johnson, enter.

The other useful situation is that you've named someone's face, or someone's voice, but not both.

- You hear that person talk to you.
  '"Bob Johnson", with a "deep, clipped, gruff" voice says, "That's my voice you're looking at, pal".

Click Bob Johnson, click "rename both", and instantly hit enter, because Bob Johnson is already in the input field. We could write code that does this automatically, though it can be a bad idea as it can be possible to abuse that when disguising yourself.

There's one more button we can add renaming dialogs: a 'rename all with this name' button. This is useful in situations where you can only see/hear either a face or a voice, but you want to rename all of them at once. For example, you've identified one of Bob Johnson's faces and two of his voices (don't ask how). You want to rename all of them to "Bob Johnson - stole eric's voice". Just examine him, click his face name, and hit the 'rename all with this name' button to change his one face name and two voice names.

All these buttons might be a little bit complicated to use, but they're out of the way, requiring a hover or click in a specific location, and can be easily learned.

### Balance considerations and pre-naming

One thing I haven't mentioned yet is that the system we've built so far can already completely emulate SS13's system. Simply make all generated faceprints and voiceprints be the real name of the character who spawned in with them at round start, and disable access to descriptions, renames, and face/voiceprint memories.

While we're on the topic of SS13, it's a good time to mention balance and how this identity system will fit with SS14's gamemodes. This is not the only attempt at a new identity system that has ever been made. A similar one was created for /TG/station, with some major similarities and some major differences (https://tgstation13.org/wiki/User:JJRcop/Identity).

The TG system crashed and burned almost as soon as it arrived, with many heralding its doom. The main reason for this was that it massively changed the balance of the game by making the crew have to learn each others' names, faces, and voices. Everyone basically being an 'Unknown' in a server of 90+ pop massively changed the meta to something unfun, the biggest example beign all the crap spammed in the common channel because no one could effectively be punished for it anyway.

Our new identity system can certainly heed the lesson taught there. For a start, we can simply have characters spawn with pre-knowledge of certain other faces and voices. NT Crew, for example, can have "pre-knowledge" of all other NT Crew - even the latejoining ones - so that they can identify other crewmembers through their faces and voices without having to put in any effort. Similarly, nuke ops can know each other without also knowing the crew. 

It can be a good idea to have all "pre-known" faces and voices hidden from the player-accessible memories screen until they are actually seen or heard. There's two major situations why this can be useful:

- An admin creates a special event character that they intend for the crew to recognize, such as a high-ranking NT executive. They can make all characters "pre-know" this person without their finding out prematurely by checking their memories.

- A non-crew antag wants to pretend to be a crewmember. This one's a little bit complicated to explain. Think of a station where all NT crew know all other NT crew, including latejoiners. In-character, this means that all crewmembers are aware of the names, faces, and voices of additional crewmembers who have yet arrive on the shuttle. A non-crew antag can pretend to be one of these yet-to-arrive people by using some tool to change their face and voice to one of these well-known backup workers. In code, their using this tool will simply add this fictional-but-ICly-real persona's face and voice to all characters' memories.

### TODO: come up with an "ending" title

Overall, it looks like our new system works. It greatly expands the possibility of unique gamemodes, servers, and challenges in SS14, strengthening roleplay, providing a more immersive experience to the player, and making faces and voices a detailed thing too. 

Perhaps the most exciting feature to come will be roundstart traitor reports issuing a 50% chance of low-browed or gravelly-voiced persons to be members of the syndicate.

**Comment by captkirk88** [comment]
You might want to read this nice article on trying to achieve reality in video game development versus what you can actually do (usually fake it).  http://cowboyprogramming.com/2007/01/05/shattering-reality/

**Comment by zamolxius** [comment]
how about man #1 says "fuck you"
man #2 says, "no you"

**Comment by digitalic** [comment]
Closing to merge into hackmd docs
